packages/flutter/lib/src/widgets/app_model.dart

```diff
@@ -0,0 +1,207 @@
+  /// is the type of the value. The value type is nullable, for example
+  /// `AppModel.setValue<String, String>(key,value)` means that the key parameter
+  /// must non-null but the value parameter can be null or a string.
+  static void setValue<K extends Object, V>(BuildContext context, K key, V? value) {
+    final _AppModelData? model = context.findAncestorWidgetOfExactType<_AppModelData>();
+    assert(_debugHasAppModel(model, context, 'setValue'));
+    model!.appModelState.setValue<K, V>(key, value);
+  }
+
+  /// Unconditionally changes the app model's `value` for `key`.
+  ///
+  /// This method should be used to lazily initialize a value that's
+  /// needed at build time, when rebuilding dependent widgets isn't
+  /// allowed or necessary. The `SharedObject` example above
+  /// demonstrates this.
+  ///
+  /// Unlike [AppModel.setValue], this method does _not_ cause dependent widgets
+  /// to be rebuilt.
+  ///
+  /// The type parameter `K` is the type of the value's keyword and `V`
+  /// is the type of the value. The value type is nullable, for example
+  /// `AppModel.initValue<String, String>(key,value)` means that the key parameter
+  /// must non-null but the value parameter can be null or a string.
+  static void initValue<K extends Object, V>(BuildContext context, K key, V? value) {
+    final _AppModelData? model = context.findAncestorWidgetOfExactType<_AppModelData>();
+    assert(_debugHasAppModel(model, context, 'initValue'));
+    model!.appModelState.initValue<K, V>(key, value);
+  }
+
+  static bool _debugHasAppModel(_AppModelData? model, BuildContext context, String methodName) {
+    assert(() {
+      if (model == null) {
+        throw FlutterError.fromParts(
+          <DiagnosticsNode>[
+            ErrorSummary('No AppModel widget found.'),
+            ErrorDescription('AppModel.$methodName requires an AppModel widget ancestor.\n'),
+            context.describeWidget('The specific widget that could not find an AppModel ancestor was'),
+            context.describeOwnershipChain('The ownership chain for the affected widget is'),
+            ErrorHint(
+              'Typically, the AppModel widget is introduced by the MaterialApp '
+              'or WidgetsApp widget at the top of your application widget tree. It '
+              'provides a key/value map of data that is shared with the entire '
+              'application.',
+            ),
+          ],
+        );
+      }
+      return true;
+    }());
+    return true;
+  }
+}
+
+class _AppModelState extends State<AppModel> {
+  late Map<Object, Object?> data = <Object, Object?>{};
+
+  @override
+  Widget build(BuildContext context) {
+    return _AppModelData(appModelState: this, child: widget.child);
+  }
+
+  V? getValue<K extends Object, V>(K key) => data[key] as V?;
+
+  void setValue<K extends Object, V>(K key, V? value) {
+    if (data[key] != value) {
+      setState(() {
+        data = Map<Object, Object?>.from(data);
```

Comment:
Won't this try to notify all consumers when a  key changes, even when the key they are listening too didn't change?

This would make notification really expensive if there's a large number of keys

Reply:
Only the consumers that depend on a key (because they used that key to lookup a value) are rebuilt.

Reply:
Only the consumers where the value changed are indeed rebuilt.

But as far as I understand, the rebuild trigger would involve call-in == on all values/keys. 

In a Riverpod application, which solves a similar problem (no umbrella widget), the application could have thousands of such keys. Including values/keys which override == to have a deep collection comparison.

Reply:
For comparison, Riverpod's notification mechanism will only call == on the changed key.

And it won't need to call == on values at all. So whether there's only one key or thousands of keys, the notification process takes the same amount of time. 
Only adding consumers increase the time it takes to notify listeners

Reply:
It creates a new Map, which will call updateShouldNotify followed but updateShouldNotifyDependents, which iterates over all the aspects that a consumer depends on

So my understanding is that this would end up calling == on all keys/values with active listeners (possibly multiple times if a key has multiple consumers)

Reply:
It's true that calling setValue() once will cause operator== on the value twice: once at the call site and once in updateShouldNotifyDependents. Since setValue() could be called more than once before dependent widgets are rebuilt, the second == check covers the possibility that the new value is no longer different than the original one. It's also true that when multiple widgets depend on the same value, the == check is repeated unnecessarily. If these proves to be a problem in practice we could complicate the implementation a little.